sdk/core/src/pipeline.rs

```diff
@@ -0,0 +1,49 @@
+use crate::policies::{Policy, PolicyResult};
+use crate::{Context, Request, Response};
+use std::sync::Arc;
+
+/// Execution pipeline.
+///
+/// A pipeline follows a precise flow:
+///
+/// 1. Per call policies are executed. Per call policies can fail and bail out of the pipeline
+///    immediately.
+/// 2. Retry policy. It allows to reexecute the following policies.
+/// 3. Per retry policies. Per retry polices are always executed at least once but are reexecuted
+///    in case of retries.
+/// 4. Transport policy. Transtport policy is always the last policy and is the policy that
+///    actually constructs the `Response` to be passed up the pipeline.
+///
+/// A pipeline is immutable. In other words a policy can either succeed and call the following
+/// policy of fail and return to the calling policy. Arbitrary policy "skip" must be avoided (but
+/// cannot be enforced by code). All policies except Transport policy can assume there is another following policy (so
+/// self.pipeline[0] is always valid).
+#[derive(Debug, Clone)]
+pub struct Pipeline {
+    pipeline: Vec<Arc<dyn Policy>>,
+}
+
+impl Pipeline {
+    pub fn new(
+        per_call_policies: Vec<Arc<dyn Policy>>,
+        retry: Arc<dyn Policy>,
```

Comment:
Why make client libraries pass in the retry and transport policies? These should added by default. See https://github.com/Azure/azure-sdk-for-net/blob/8d19a820660c3c34b6ebe7904bd4e7ff895c82f6/sdk/core/Azure.Core/src/Pipeline/HttpPipelineBuilder.cs#L35. This is fairly consistent across languages. If an SDK doesn't want a retry, they can turn it off via ClientOptions.

Reply:
Let's make that change after we merge this PR.